runtime/jvm/src/main/kotlin/kotlinx/serialization/Resolving.kt

```diff
@@ -47,38 +48,76 @@ private fun mapJavaClassNameToKotlin(s: String): String = when(s) {
+fun serializerByTypeToken(type: Type): KSerializer<Any> {
+    return when (type) {
+        is GenericArrayType -> {
+            val eType = type.genericComponentType.let {
+                when (it) {
+                    is WildcardType -> it.upperBounds.first()
+                    else -> it
+                }
+            }
+            val serializer = serializerByTypeToken(eType)
+            val kclass = when (eType) {
+                is ParameterizedType -> (eType.rawType as Class<*>).kotlin
+                is KClass<*> -> eType
+                else -> throw IllegalStateException("unsupported type in GenericArray: ${eType::class}")
+            } as KClass<Any>
+            ReferenceArraySerializer<Any, Any>(kclass, serializer) as KSerializer<Any>
+        }
+        is Class<*> -> if (!type.isArray) {
+            serializerByClass(type.kotlin)
+        } else {
+            val eType: Class<*> = type.componentType
+            val s = serializerByTypeToken(eType)
+            val arraySerializer = ReferenceArraySerializer<Any, Any>(eType.kotlin as KClass<Any>, s)
+            arraySerializer as KSerializer<Any>
+        }
+        is ParameterizedType -> {
+            val rootClass = (type.rawType as Class<*>)
+            val args = (type.actualTypeArguments)
+            when {
+                List::class.java.isAssignableFrom(rootClass) -> ArrayListSerializer(serializerByTypeToken(args[0])) as KSerializer<Any>
+                Set::class.java.isAssignableFrom(rootClass) -> HashSetSerializer(serializerByTypeToken(args[0])) as KSerializer<Any>
+                Map::class.java.isAssignableFrom(rootClass) -> HashMapSerializer(
+                    serializerByTypeToken(args[0]),
+                    serializerByTypeToken(args[1])
+                ) as KSerializer<Any>
+                Map.Entry::class.java.isAssignableFrom(rootClass) -> MapEntrySerializer(
+                    serializerByTypeToken(args[0]),
+                    serializerByTypeToken(args[1])
+                ) as KSerializer<Any>
 
-            else -> {
-                val varargs = args.map { serializerByTypeToken(it) }.toTypedArray()
-                (rootClass.invokeSerializerGetter(*varargs) as? KSerializer<Any>) ?: serializerByClass<Any>(rootClass.kotlin)
+                else -> {
+                    val varargs = args.map { serializerByTypeToken(it) }.toTypedArray()
+                    (rootClass.invokeSerializerGetter(*varargs) as? KSerializer<Any>) ?: serializerByClass<Any>(
+                        rootClass.kotlin
+                    )
+                }
             }
         }
+        is WildcardType -> serializerByTypeToken(type.upperBounds.first())
+        else -> throw IllegalArgumentException("typeToken should be instance of Class<?> or ParametrizedType, actual type is $type ${type::class}")
```

Comment:
Exception text is a bit obsolete now because you've added more clauses. It can look like `"typeToken should be an instance of Class<?>, GenericArray, ParametrizedType or WildcardType, but actual type is $type ${type::class}"`